### PR TITLE
HYPERSCALE RETURNS!

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request
image
HYPERSCALE RETURNS!!

This re-implements the old hyperscale system in a new and improved manner. Gone are the days of slicing sprites up into a million in order to apply hyperscale, now all that is needed is a sprite with a clear defined palette. (It does not even need to be greyscale! (Although you probably want it to be since it gets converted into greyscale for sorting so it's best for interpretation to sprite in greyscale.))
Just look at these sprite sheets!!
image
image

The image does not show all the hyperscale items, so I will list them:

- Jaeger MK1
- Jaeger MK2
- Style line hats
- Style line duster
- Capes and Kama
- Xenonaut Gear
- Robot armor
- All visors including Robot and the heavy xenonaut helmet

Codewise:
Hyperscale works under the greyscale system and generates at runtime.
This is the process of generation:
1) At Subsystem setup, when the greyscale configs are all initialized, the hyperscale sprites colors are gathered and sorted by darkest to lightest greyscale values, this color list is stored on the config.
2) When an armor piece is colored, a byond builtin to swap the colors is called and the existing colors are swapped out with the palette. This is then passed like normal through the system.
The most expensive part of the system only happens at game init, and has very little impact on even my machine. I think it will be fine.

I have also added a utility "prefix" var on greyscale configs to reduce icon file and config bloat. Will probably add more.

## Why It's Good For The Game
image
Doesn't this just look cool??

## Todo

- [x] Fix Loadouts (This will probably mean a wipe of old greyscale.)
- [x] Fix up any odd sprites, particularly capes. Kuro will probably help me with this.
- [x] Documentation and cleanup.
- [x] Testing and fixes.

This brings the best of both worlds. It allows easy addition of colors. Colors are curated and palettes. Sprites are maintainable as they are simply greyscale.

## Changelog
:cl:
add: Hyperscale system added to greyscale.
add: Hyperscales Jaeger Mk1, Mk2, Robot armor, Style line, Xenonaut.
/:cl:

### CODE REVIEW

**Comment by Barnet2** [comment]
We are so back.

**Comment by cazuccmuss** [comment]
speedmerge...

**Comment by zzurge** [comment]
MY DRIP GAME IS ABOUT TO COME BACK

**Comment by DorfGuy1** [comment]
the robots still don'tlook as good as they used to

**Review by Kuro020 — APPROVED** [review]
The suffering is over, the golden age begins now

**Comment by Alexeivichi** [comment]
We never left.

**Comment by novaepee** [comment]
One of hyper's many projects. Stay tune for more.

**Review by Lumipharon — APPROVED** [review]
Seems OK at first look

**Comment by Guidesu** [comment]
Right clicking on the M10XE with the painting tools does nothing.
The helmet on M10XE loses storage aswell.
Could you somehow make the modules like helmet visor like TYR and mirmir be colorable so it can fit the color scheme of a armour?

**Comment by Guidesu** [comment]
image
Looks ugly like this

**Comment by Weedmaster77** [comment]
The armor changes are nice but why cant we still have the old cloak coloring system back it was far better and made it easier to match clothing colors an for only one type of item don't think its that much of a problem to leave it as was.

**Comment by 1kajowsmol** [comment]
um scout helmet is invisible, and visors cant be colored......... 
Screenshot 2023-06-21 230019

**Comment by ilikesaying** [comment]
Round ID 21787
Screenshot 2023-06-22 000602
Screenshot 2023-06-22 000622

there's supposed to be a EOD jaeger exoskeleton here
Screenshot 2023-06-22 000712

**Comment by Tailiik** [comment]
>Remove hyperscale because it's too hard to maintain 
>Make limited colors because muh art direction consistency or something
>Bring back hyperscale because actually lol lmao it isn't too hard to maintain
>But also nerf cape colors 
>And also why test pr let's make armor invisible all day 
>Honk

In all seriousness I dunno if it's because the pr in general is broken or what but you can't recolor visors now, the one part of helmets that actually do show up, honk

**Comment by hyper2snyper** [comment]
> > Remove hyperscale because it's too hard to maintain
> > Make limited colors because muh art direction consistency or something
> > Bring back hyperscale because actually lol lmao it isn't too hard to maintain
> > But also nerf cape colors
> > And also why test pr let's make armor invisible all day
> > Honk
> 
> In all seriousness I dunno if it's because the pr in general is broken or what but you can't recolor visors now, the one part of helmets that actually do show up, honk

The visors should be colorable. Theres something really weird going on.

**Comment by Kuro020** [comment]
> > Remove hyperscale because it's too hard to maintain
> > Make limited colors because muh art direction consistency or something
> > Bring back hyperscale because actually lol lmao it isn't too hard to maintain
> > But also nerf cape colors
> > And also why test pr let's make armor invisible all day
> > Honk
> 
> In all seriousness I dunno if it's because the pr in general is broken or what but you can't recolor visors now, the one part of helmets that actually do show up, honk

maybe, its easier to maintain now because hyper redid the system and it isnt cbt to sprite it anymore? imagine that

**Comment by Pixelknightly** [comment]
> > Remove hyperscale because it's too hard to maintain
> > Make limited colors because muh art direction consistency or something
> > Bring back hyperscale because actually lol lmao it isn't too hard to maintain
> > But also nerf cape colors
> > And also why test pr let's make armor invisible all day
> > Honk
> 
> In all seriousness I dunno if it's because the pr in general is broken or what but you can't recolor visors now, the one part of helmets that actually do show up, honk

This IS the pr being tested. That's why it's called a "testmerge"
Also what kuro said.

Did you even bother to read the PR description? Or atleast familiarize yourself with what the tags mean?
Is that why you're honking a clown horn?

**Comment by Neskah** [comment]
The pseudo-shading stuff is cool, but this is a straight downgrade from the less restrictive color picker system, IMHO. Can't do little nuanced things like color your visor to match svalinn energy shields, etc, or color both elements of a cape the same shade as to have it look like it doesn't have a pattern. Things looked cool before, and now they're just ever-so-slightly different looking with significantly more restrictions. If this is an implementation strictly to eliminate our horde of femboy medics that's the most rational argument I can think of. (Or if it was somehow implemented as an optional system.)

Also, robot helmets have no visor or storage, and robot body armor's sprite does not have visual upgrades when held in the hand.

**Comment by Sirsnek15** [comment]
> The pseudo-shading stuff is cool, but this is a straight downgrade from the less restrictive color picker system, IMHO. Can't do little nuanced things like color your visor to match svalinn energy shields, etc, or color both elements of a cape the same shade as to have it look like it doesn't have a pattern. Things looked cool before, and now they're just ever-so-slightly different looking with significantly more restrictions. Also, robot helmets have no visor or storage, and robot body armor's sprite does not have visual upgrades when held in the hand.

literally everything listed here is a bug, not intended, there are no reasons or wants for these things to be changed in this way. Its not a downgrade, just a VERY bugged upgrade that needs fixing

File: code/__DEFINES/loadout.dm [hunk 1]
```diff
@@ -744,6 +744,6 @@ GLOBAL_LIST_INIT(bypass_storage_content_save, typecacheof(list(
 #define MAXIMUM_LOADOUT 50
 
 /// The current loadout version
-#define CURRENT_LOADOUT_VERSION 11
+#define CURRENT_LOADOUT_VERSION 13
 
 GLOBAL_LIST_INIT(accepted_loadout_versions, list(5, 6, 7, 8, 9, 10, 11))
```
Comment (Lumipharon) [inline code/__DEFINES/loadout.dm:749]:
```suggestion
GLOBAL_LIST_INIT(accepted_loadout_versions, list(5, 6, 7, 8, 9, 10, 11, 13))
```
Needed for loadout imports.

**Comment by Neskah** [comment]
> 

Oh, fair. Are we intended to be able to use full RGB coloration on cloaks/visors?

**Comment by hyper2snyper** [comment]
> > 
> 
> Oh, fair. Are we intended to be able to use full RGB coloration on cloaks/visors?

Alright so there seems to be some miscommunication here. Everything about your first comment is actually ok. None are bugs, the color wheel is gone.

So the current issue of regular sprites is that maintenance and multiple colors are exponential, the walls of sprites should attest to that. The issue with greyscale is that you lose out on all manner of color shifting. Yellows, reds and oranges all look terrible. However it's capable of many many colors. Hyperscale is the happy medium.

We used to have hyperscale about a year ago, it was always meant to replace both color wheel and manual presets. However, it was very hard to sprite for and byond issues caused it to be deleted for some time. (I also had a lot of life stuff so I couldn't work on it)

Color wheel is gone so that better looking palettes can instead be used.

However, this is only the first in my planned color and drip related prs. The next one should expand this system and make it easier to spread. Depending on permission (if I can convince kuro) and implementation (if I can get it to work in the code) I do plan on allowing both color wheel and palettes for secondary colors such as visors and cape trims. For cape trims I also plan to add a bunch of variants including the ability to remove it. 

I hope this cleared some stuff up.

File: code/datums/greyscale/layer.dm [hunk 2]
```diff
@@ -64,9 +72,62 @@
...
+	var/icon_state = json_data["icon_state"]
+	if(prefix && !json_data["ignore_prefix"])
+		icon_state = prefix+icon_state
+	if(!(icon_state in icon_states(icon_file)))
+		CRASH("Configured icon state \[[icon_state]\] was not found in [icon_file]. Double check your json configuration.")
+	icon = new(icon_file, icon_state)
+	icon.GrayScale()
+	for(var/dir in list(SOUTH, NORTH, EAST, WEST, NORTHWEST, NORTHEAST, SOUTHWEST, SOUTHEAST))
+		for(var/x=0, x<icon.Width(), x++)
+			for(var/y=0, y<icon.Height(), y++)
```
Comment (TiviPlus) [inline code/datums/greyscale/layer.dm:108]:
```suggestion
		for(var/x=0 to icon.Width())
			for(var/y=0 to icon.Height())
```
might need --icon.Height()

File: code/datums/greyscale/layer.dm [hunk 3]
```diff
@@ -64,9 +72,62 @@
...
+				var/pixel = icon.GetPixel(x,y, dir=dir)
+				if(!pixel || (pixel in icon_file_colors)) continue;
```
Comment (TiviPlus) [inline code/datums/greyscale/layer.dm:110]:
```suggestion
				if(!pixel || (pixel in icon_file_colors))
					continue
```

File: code/datums/greyscale/layer.dm [hunk 4]
```diff
@@ -64,9 +72,62 @@
...
+				var/rgb_val = ReadRGB(pixel)[1]
+				if(!length(icon_file_colors))
+					icon_file_colors.Add(pixel)
+					continue
+				for(var/i=1, i<=length(icon_file_colors), i++)
```
Comment (TiviPlus) [inline code/datums/greyscale/layer.dm:115]: Ditto use byond format

File: code/datums/greyscale/layer.dm [hunk 5]
```diff
@@ -64,9 +72,62 @@
...
+				if(pixel in icon_file_colors)
+					continue
+				icon_file_colors.Add(pixel)
+
+/datum/greyscale_layer/hyperscale/InternalGenerate(list/colors, list/render_steps)
+	. = ..()
+	var/icon/new_icon = icon(icon)
+	if(length(icon_file_colors) > length(colors))
+		CRASH("[src] set to Hyperscale, expected [length(icon_file_colors)], got [length(colors)].")
+	for(var/i=1, i<=length(icon_file_colors), i++)
```
Comment (TiviPlus) [inline code/datums/greyscale/layer.dm:130]: Ditto use byond format

File: code/datums/loadout/loadout_manager.dm [hunk 6]
```diff
@@ -159,6 +159,24 @@
...
+			module.colors = initial(module.item_type.greyscale_colors)
+			update_attachments(module.attachments, version)
+			if(!istype(module, /datum/item_representation/armor_module/armor) && !istype(module, /datum/item_representation/armor_module/colored))
+				continue
+			var/datum/item_representation/armor_module/new_module = new
+			new_module.attachments = module.attachments
+			new_module.item_type = module.item_type
+			new_module.colors = module.colors
+			attachments.Remove(module)
+			attachments.Add(new_module)
```
Comment (TiviPlus) [inline code/datums/loadout/loadout_manager.dm:177]: dunno why this is in a seperate proc?

File: code/modules/clothing/modular_armor/attachments.dm [hunk 7]
```diff
@@ -272,3 +236,15 @@
...
+///Colors the armor when the parent is right clicked with facepaint.
+/obj/item/armor_module/armor/proc/handle_color(datum/source, obj/I, mob/user)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, TYPE_PROC_REF(/atom, attackby), I, user)
+	return COMPONENT_NO_AFTERATTACK
+
+///Relays the extra controls to the user when the parent is examined.
+/obj/item/armor_module/armor/proc/extra_examine(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+	examine_list += "Right click the [parent] with paint to color the [src]"
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/attachments.dm:249]:
```suggestion
	examine_list += "Right click the [parent] with paint to color [src]"
```

File: code/modules/clothing/modular_armor/modular.dm [hunk 8]
```diff
@@ -326,46 +398,107 @@
...
-	if(!do_after(user, 1 SECONDS, TRUE, src, BUSY_ICON_GENERIC))
+	var/selection
+	switch(colorable_allowed)
+		if(COLOR_WHEEL_ONLY)
+			selection = "Color Wheel"
+		if(COLOR_WHEEL_ALLOWED)
+			selection = list("Color Wheel", "Preset Colors")
+			selection = tgui_input_list(user, "Choose a color setting", "Choose setting", selection)
+		if(COLOR_WHEEL_NOT_ALLOWED)
+			selection = "Preset Colors"
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/modular.dm:474]: Use defines for these strings please

File: code/modules/clothing/modular_armor/modular.dm [hunk 9]
```diff
@@ -413,45 +549,103 @@
...
+
```
Comment (TiviPlus) [inline code/modules/clothing/modular_armor/modular.dm:623]: Ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Impressively simple solution very nice

**Comment by Lumipharon** [comment]
Causes #13351 
Was fixed in the jaeger rework.

**Review by Lumipharon — APPROVED** [review]

**Pull request merged**